File: Bronze/2018/US Open/03_family_tree/family_tree_c.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class TreeNode {
public:
    string name;
    TreeNode* mother;
    vector<TreeNode*> children;

    TreeNode(string name) : name(name), mother(nullptr) {}
};

int getAncestorDistance(TreeNode* sourceCow, TreeNode* targetCow) {
    int distance = 0;
    TreeNode* node = sourceCow;
    while (node != nullptr) {
        if (node == targetCow) {
            return distance;
        }
        node = node->mother;
        distance++;
    }
    return -1;
}

int main() {
    ifstream fin("family.in");
    ofstream fout("family.out");

    int N;
    fin >> N;
    string name_a, name_b;
    fin >> name_a >> name_b;

    unordered_map<string, TreeNode*> graph;
    for (int i = 0; i < N; ++i) {
        string name_i, name_j;
        fin >> name_i >> name_j;

        if (graph.find(name_i) == graph.end()) {
            graph[name_i] = new TreeNode(name_i);
        }
        if (graph.find(name_j) == graph.end()) {
            graph[name_j] = new TreeNode(name_j);
        }

        TreeNode* node_i = graph[name_i];
        TreeNode* node_j = graph[name_j];

        node_i->children.push_back(node_j);
        node_j->mother = node_i;
    }

    TreeNode* node = graph[name_a];
    int da = 0;
    while (node != nullptr) {
        if (getAncestorDistance(graph[name_b], node) != -1) {
            break;
        }
        node = node->mother;
        da++;
    }

    // cannot find a common ancestor between a and b
    if (node == nullptr) {
        fout << "NOT RELATED";
        return 0;
    }

    int db = getAncestorDistance(graph[name_b], node);
    if (da > 1 && db > 1) {
        fout << "COUSINS";
        return 0;
    }
    if (da == 1 && db == 1) {
        fout << "SIBLINGS";
        return 0;
    }

    // find a node with higher generation
    if (da > db) {
        swap(da, db);
        swap(name_a, name_b);
    }

    string prefix = "";
    for (int i = 0; i < max(0, db - 2); ++i) {
        prefix += "great-";
    }
    if (db > 1 && da == 0) {
        prefix += "grand-";
    }
    if (da == 0) {
        prefix += "mother";
    } else {
        prefix += "aunt";
    }

    fout << name_a << " is the " << prefix << " of " << name_b;
    return 0;
}

File: Bronze/2018/US Open/03_family_tree/FamilyTreeC.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.*;

/**
 * @author Red
 * @version 1.0
 */
public class FamilyTreeC {
    static int N;
    static Map<String, TreeNode> graph = new HashMap<>();

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("family.in"));
        PrintWriter pw = new PrintWriter("family.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        N = Integer.parseInt(st.nextToken());
        String name_a = st.nextToken();
        String name_b = st.nextToken();
        for (int i = 0; i < N; i++) {
            String[] names = r.readLine().trim().split(" ");
            String name_i = names[0];
            String name_j = names[1];

            if (!graph.containsKey(name_i)) {
                graph.put(name_i, new TreeNode(name_i));
            }
            if (!graph.containsKey(name_j)) {
                graph.put(name_j, new TreeNode(name_j));
            }

            TreeNode node_i = graph.get(name_i);
            TreeNode node_j = graph.get(name_j);

            node_i.children.add(node_j);
            node_j.mother = node_i;
        }

        TreeNode node = graph.get(name_a);
        int da = 0;
        while (node != null) {
            if (getAncestorDistance(graph.get(name_b), node) != -1) {
                break;
            }
            node = node.mother;
            da++;
        }

        // cannot rd a common ancestor between a and b
        if (node == null) {
            pw.write("NOT RELATED");
            pw.close();
            System.exit(0);
        }

        int db = getAncestorDistance(graph.get(name_b), node);
        if (da > 1 && db > 1) {
            pw.write("COUSINS");
            pw.close();
            System.exit(0);
        }
        if (da == 1 && db == 1) {
            pw.write("SIBLINGS");
            pw.close();
            System.exit(0);
        }

        // rd a node with higher generation
        if (da > db) {
            int temp = da;
            da = db;
            db = temp;
            String tempName = name_a;
            name_a = name_b;
            name_b = tempName;
        }

        StringBuilder prefix = new StringBuilder("great-".repeat(Math.max(0, db - 2)));
        if (db > 1 && da == 0) {
            prefix.append("grand-");
        }
        if (da == 0) {
            prefix.append("mother");
        } else {
            prefix.append("aunt");
        }

        pw.write(name_a + " is the " + prefix + " of " + name_b);
        pw.close();
    }

    /**
     * 返回 sourceCow 到 targetCow 的辈份距离
     */
    static int getAncestorDistance(TreeNode sourceCow, TreeNode targetCow) {
        int distance = 0;
        TreeNode node = sourceCow;
        while (node != null) {
            if (node == targetCow) {
                return distance;
            }
            node = node.mother;
            distance++;
        }
        return -1;
    }


}

class TreeNode {
    String name;
    TreeNode mother;
    List<TreeNode> children = new ArrayList<>();
    TreeNode(String name) {
        this.name = name;
    }
}
